viewer-prototype/src/node/trace-server-service.ts

```diff
@@ -1,33 +1,51 @@
-import { spawn, exec } from 'child_process';
+import { spawn } from 'child_process';
 import { injectable } from 'inversify';
-import { TraceServerConfigService } from '../common/trace-server-config';
+import { PortBusy, TraceServerConfigService } from '../common/trace-server-config';
+import treeKill = require('tree-kill');
 @injectable()
 export class TraceServerServiceImpl implements TraceServerConfigService {
+    private processId: number;
 
-    startTraceServer(path: string | undefined, port: number | undefined): Promise<void> {
+    startTraceServer(path: string | undefined, port: number | undefined): Promise<string> {
         const server = spawn(`${path}`, ['-vmargs', `-Dtraceserver.port=${port}`]);
-        return new Promise<void>((resolve, reject) => {
+        this.processId = server.pid;
+        return new Promise<string>((resolve, reject) => {
+            // TODO: Since the resolve never actually gets returned, we need a better way to determine if the child process
+            // initiated by the spawn successfully starts the trace server
+
             // If the server provides output on any channel before it exits, consider that a success.
             // That doesn't mean that the server has really started. On the frontend, the `TraceServerConnectionStatusService`
             // will ping the port until it receives a response, which is the official measure of success.
-            server.stdout.on('data', () => resolve());
-            server.stderr.on('data', () => resolve());
-            // If the server exits or errors before it outputs, consider it a failure.
-            server.on('exit', code => reject(code));
-            server.on('error', error => reject(error));
-        })
-            .finally(() => { server.removeAllListeners(); });
+
+            server.stdout.on('data', () => setTimeout(()=> resolve('success'), 2000));
+            server.stderr.on('data', () => setTimeout(()=> resolve('success'), 2000));
+             // If the server exits or errors before it outputs, consider it a failure.
+            server.on('error', err => {
+                reject(err);
+            });
+            server.on('exit', code => {
+                reject(PortBusy(code));
```

Comment:
maybe code 0 can be seen as success as well. Non-zero code values are for the reject then.